[Blog] A Deep-Dive into Flink's Network Stack (#218)
apache/flink-web: _posts/2019-06-04-flink-network-stack.md

# Logical View

Flink’s network stack provides the following logical view to the subtasks when communicating with each other, for example during a network shuffle as required by a `keyBy()`.

<center>
<img src="{{ site.baseurl }}/img/blog/2019-06-04-network-stack/flink-network-stack1.png" width="400px" alt="Logical View on Flink's Network Stack"/>
</center>
<br>

It abstracts over the different settings of the following three concepts:

* Subtask output type (`ResultPartitionType`):
    * **pipelined (bounded or unbounded):**
    Sending data downstream as soon as it is produced, potentially one-by-one, either as a bounded or unbounded stream of records.
    * **blocking:**
    Sending data downstream only when the full result was produced.

* Scheduling type:
    * **all at once (eager):**
    Deploy all subtasks of the job at the same time (for streaming applications).
    * **next stage on first output (lazy):**
    Deploy downstream tasks as soon as any of their producers generated output.
    * **next stage on complete output:**
    Deploy downstream tasks when any or all of their producers have generated their full output set.

* Transport:
    * **high throughput:**

>     Sending each record one-by-one obviously does not give the best throughput. For this, Flink has to buffer a bunch of records into its network buffers and send them altogether.

Reviewer: ```suggestion
    Instead of sending each record one-by-one, Flink buffers a bunch of records into its network buffers and send them altogether. This reduces the cost per record and leads to high throughput.
```

Reviewer: Just a suggestion :)

    #channels * buffers-per-channel + floating-buffers-per-gate

The total number of buffers on a single TaskManager usually does not need configuration. See the [Configuring the Network Buffers](https://ci.apache.org/projects/flink/flink-docs-release-1.8/ops/config.html#configuring-the-network-buffers) documentation for details on how to do so if needed.

Whenever a subtask’s sending buffer pool is exhausted — buffers reside in either a result subpartition's buffer queue or inside the lower, Netty-backed network stack — the producer is blocked, cannot continue, and experiences backpressure. The receiver works in a similar fashion: any incoming Netty buffer in the lower network stack needs to be made available to Flink via a network buffer. If there is no network buffer available in the appropriate subtask's buffer pool, Flink will stop reading from this channel until a buffer becomes available. This would effectively backpressure all sending subtasks on this multiplex:
<br>

<center>

> <img src="{{ site.baseurl }}/img/blog/2019-06-04-network-stack/flink-network-stack3.png" width="700px" alt="Physical-transport-backpressure-Flink's Network Stack"/>

Reviewer: Is B.4 is causing the backpressure?

Author: I clarified this a bit better now / added a description of what the picture is showing

<br>

# Latency vs. Throughput

Network buffers were introduced to get higher resource utilisation and higher throughput at the cost of having some records wait in buffers a little longer. Although an upper limit to this wait time can be given by the buffer timeout, you may be curious to find out more about the trade-off between these two dimensions: latency and throughput, as, obviously, you cannot get both. The following plot shows various values for the buffer timeout starting at 0 (flush with every record) to 100ms (the default) and shows the resulting throughput rates on a cluster with 100 nodes and 8 slots each running a job that has no business logic and thus only tests the network stack. For comparison, we also plot Flink 1.4 before the low-latency improvements (as described above) were added.
<br>

<center>
<img src="{{ site.baseurl }}/img/blog/2019-06-04-network-stack/flink-network-stack9.png" width="650px" alt="Network-buffertimeout-Flink's Network Stack"/>
</center>

> <br>

Reviewer: it might make sense to write down the results from the benchmark as well.
People might not get what the picture wants to tell them. (and I see a risk of people just scrolling through the pictures, drawing conclusions themselves)

<br>

<center>
<img src="{{ site.baseurl }}/img/blog/2019-06-04-network-stack/flink-network-stack4.png" width="700px" alt="Physical-transport-credit-flow-Flink's Network Stack"/>
</center>
<br>

Credit-based flow control will use [buffers-per-channel](https://ci.apache.org/projects/flink/flink-docs-release-1.8/ops/config.html#taskmanager-network-memory-buffers-per-channel) to specify how many buffers are exclusive (mandatory) and [floating-buffers-per-gate](https://ci.apache.org/projects/flink/flink-docs-release-1.8/ops/config.html#taskmanager-network-memory-floating-buffers-per-gate) for the local buffer pool (optional<sup>3</sup>) thus achieving the same buffer limit as without flow control. The default values for these two parameters have been chosen so that the maximum (theoretical) throughput with flow control is at least as good as without flow control, given a healthy network with usual latencies. You may need to adjust these depending on your actual round-trip-time and bandwidth.
<br>

<sup>3</sup>If there are not enough buffers available, each buffer pool will get the same share of the globally available ones (± 1).

## Inflicting Backpressure (2)

> As opposed to the receiver's backpressure mechanisms without flow control, credits provide a more direct control: If a receiver cannot keep up, its available credits will eventually hit 0 and stop the sender from forwarding buffers to the lower network stack. There is backpressure on this logical channel only and no need to block reading from a multiplexed TCP channel. Other receivers are therefore not affected in processing available buffers.

Reviewer 2: ```suggestion
As opposed to the receiver's backpressure mechanisms without flow control, credits provide a more direct control: If a receiver cannot keep up, its available credits will eventually hit 0 and stop the sender from forwarding buffers to the lower network stack. There is backpressure on this logical channel only and there is no need to block reading from a multiplexed TCP channel. Other receivers are therefore not affected in processing available buffers.
```

## What do we Gain? Where is the Catch?

<img align="right" src="{{ site.baseurl }}/img/blog/2019-06-04-network-stack/flink-network-stack5.png" width="300" height="200" alt="Physical-transport-credit-flow-checkpoints-Flink's Network Stack"/>

> Since, with flow control, a channel in a multiplex can not block another of its logical channels, the overall resource utilisation should increase. In addition, by having full control over how much data is “on the wire”, we are also able to improve [checkpoint alignments](https://ci.apache.org/projects/flink/flink-docs-release-1.8/internals/stream_checkpointing.html#checkpointing): without flow control, it would take a while for the channel to fill the network stack’s internal buffers and propagate that the receiver is not reading anymore. During that time, a lot of buffers could be sitting around. Any checkpoint barrier would have to queue up behind these buffers and would thus have to wait until all of those have been processed before it can start (“Barriers never overtake records!”).

Reviewer 2: ```suggestion
Since, with flow control, a channel in a multiplex cannot block another of its logical channels, the overall resource utilisation should increase. In addition, by having full control over how much data is “on the wire”, we are also able to improve [checkpoint alignments](https://ci.apache.org/projects/flink/flink-docs-release-1.8/internals/stream_checkpointing.html#checkpointing): without flow control, it would take a while for the channel to fill the network stack’s internal buffers and propagate that the receiver is not reading anymore. During that time, a lot of buffers could be sitting around. Any checkpoint barrier would have to queue up behind these buffers and would thus have to wait until all of those have been processed before it can start (“Barriers never overtake records!”).
```
